Segment resume API

saurabhd336 opened the pull request:
This PR adds controller API support for resuming consumption for realtime tables after deletion of CONSUMING segment from ZK. Allows passing offsets to resume reading from per partition.

GH issue: https://github.com/apache/pinot/issues/6679

saurabhd336 commented:
cc: @npawar @Jackie-Jiang

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+  public String resumeRealtimeTable(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      RealtimeSegmentResumeConfig realtimeSegmentResumeConfig) {
+    TableConfig tableConfig = _pinotHelixResourceManager.getTableConfig(tableName, TableType.REALTIME);
```
Let's move the actual logic into the `PinotLLCRealtimeSegmentManager` and inject that into this API class.

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+
+  // TODO add support for choosing default offset for other partitions
+  private static class RealtimeSegmentResumeConfig {
+    private Map<Integer, Long> _partitionIdToOffsetMap;
```
We should not ask user to provide these info. As a starting point, we can use the end offset of the last completed segment as the start offset of the new segment. Then we can add support to all the `OffsetCriteria` and also the custom offset

saurabhd336 replied:
I was following the discussion on https://github.com/apache/pinot/issues/6679, and the general consensus seemed to be on first allowing explicitly specifying partitionId -> offset to create new consuming segments, and then support implicitly deriving the offsets from existing segments?
If that's not the case, I will change this

saurabhd336 replied:
Anyway I've made the changes. We can add support for explicitly specifying offsets on top of that.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+      long newSegmentCreationTimeMs = System.currentTimeMillis();
+
+      // TODO(saurabh) Get the sequence number from ideal state (largest seq # + 1)
+      int seqNumber = Math.abs(new Random().nextInt());
```
This to-do should be addressed in the first version

saurabhd336 replied:
Yes thats the plan. Will address

saurabhd336 replied:
Addressed

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+
+    for (Map.Entry<Integer, Long> partitionOffset : realtimeSegmentResumeConfig.getPartitionIdToOffsetMap()
+        .entrySet()) {
+      int partitionId = partitionOffset.getKey();
```
First check if there is already a consuming segment for this partition

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+  @Path("/segmentState/{tableName}/resumeRealtimeTable")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Resume a realtime table", notes = "Resume a segment")
```
```suggestion
  @ApiOperation(value = "Resume the consumption of a realtime table", notes = "Resume the consumption of a realtime table")
```

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+
+
+@Api(tags = Constants.SEGMENT_TAG)
```
I feel this should be under the table tag

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  @POST
+  @Path("/segmentState/{tableName}/resumeRealtimeTable")
```
```suggestion
  @Path("/tables/{tableName}/resumeConsumption")
```

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,144 @@
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Resume a realtime table", notes = "Resume a segment")
+  public String resumeRealtimeTable(
```
It can take an optional parameter of the partitions to be resumed

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -701,7 +702,7 @@ private void createNewSegmentZKMetadata(TableConfig tableConfig, PartitionLevelS
 
   @Nullable
-  private SegmentPartitionMetadata getPartitionMetadataFromTableConfig(TableConfig tableConfig, int partitionId) {
+  public static SegmentPartitionMetadata getPartitionMetadataFromTableConfig(TableConfig tableConfig, int partitionId) {
```
(minor) Not needed

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1266,7 +1267,7 @@ long getCurrentTimeMs() {
   }
 
-  private int getNumReplicas(TableConfig tableConfig, InstancePartitions instancePartitions) {
+  public static int getNumReplicas(TableConfig tableConfig, InstancePartitions instancePartitions) {
```
(minor) Not needed

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1385,4 +1386,78 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
+   * Creates PROPERTYSTORE and IDEALSTATE entries for each partitionGroup to resume realtime table consumption
+   * @param tableName
+   */
+  public void resumeRealtimeTableConsumption(String tableName) {
```
Please take a look at `ensureAllPartitionsConsuming()`, the underlying logic should be similar, and we should be able to reuse some code

saurabhd336 replied:
Ack. Refactored to add the functionality in `ensureAllPartitionsConsuming()` itself.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1385,4 +1386,78 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
+    Map<Integer, SegmentZKMetadata> partitionToLatestSegment = getLatestSegmentZKMetadataMap(realtimeTableName);
+    for (PartitionGroupMetadata partitionGroupMetadata : newPartitionGroupMetadataList) {
+      int partitionId = partitionGroupMetadata.getPartitionGroupId();
+      if (partitionToLatestSegment.containsKey(partitionId) && Status.IN_PROGRESS.equals(
```
Safer to use ideal state to identify whether the segment is consuming. Segment might be `OFFLINE` even though its status is `IN_PROGRESS`

saurabhd336 replied:
Ack. Getting it from idealState now. Checking for CONSUMING segments

Jackie-Jiang reviewed (commented):
Add @sajjad-moradi to also take a look

sajjad-moradi reviewed (commented):
There is a [design doc](https://docs.google.com/document/d/1uuXuif0SfMTnH3ykfuzXXNjh-LyszjilodD_mdlilZg/edit#heading=h.tvfsvrm5pwew) for a broader topic "pause/resume consumption on realtime table". The functionality of the endpoint provided in this PR will be part of the resume endpoint mentioned in the design doc.

If there's not an urgent need to have this functionality, I'm going to work on the pause/resume soon so we don't need to refactor the code in this PR when pause/resume is implemented. 

But if this functionality is needed ASAP, for now we can have it in the resume endpoint and later on when we implement the broader pause/resume feature, we should make sure that the resume endpoint works for paused tables and it also support this PR's functionality. We should be careful for the API to be forward compatible with pause/resume.

saurabhd336 commented:
@sajjad-moradi I had a discussion with @npawar on this, and seems like this functionality addresses issues arising due to accidental deletion of consuming segments, and therefore its availability is of importance, beyond the Pause / Resume feature too.
From the Pause / Resume design doc's perspective, I see this API fitting the Resume API description fairly well and the "in one Ideal State update operation, sets the “isTablePaused” flag to false" operation can be added to this, optionally if the table is indeed paused? Do you see any challenges with that? Are we saying we'd need to write the resume API from scratch when implementing the Pause / Resume feature?

cc: @npawar for comments

sajjad-moradi commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1385,4 +1386,89 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
+            : new LLCSegmentName(latestSegmentMeta.getSegmentName()).getSequenceNumber() + 1;
+
+        LLCSegmentName newLLCSegment = new LLCSegmentName(tableName, partitionId, seqNumber, newSegmentCreationTimeMs);
+        SegmentZKMetadata newSegmentZKMetadata = new SegmentZKMetadata(newLLCSegment.getSegmentName());
+        newSegmentZKMetadata.setCreationTime(newLLCSegment.getCreationTimeMs());
+        newSegmentZKMetadata.setStartOffset(offset);
+        newSegmentZKMetadata.setNumReplicas(numReplicas);
```
`createNewSegmentZKMetadata` can be used here.

sajjad-moradi reviewed (commented):
It's not easy to add a unit test here. Could you please add a small section to the PR description on how you tested the new functionality.

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1385,4 +1386,89 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
+            getMaxNumPartitionsPerInstance(instancePartitions, numPartitionGroups, numReplicas),
+            newPartitionGroupMetadataList);
+
+        _helixResourceManager.getPropertyStore().set(
```
we have to ensure that this change plays nicely with the code in ensureAllPartitionsConsuming. That method also creates new zkMetadata and idealState entry, and these 2 should not interfere with each other.  You'll have to add some synchronization on table name between the 2 methods

saurabhd336 replied:
With the redesign, this is achieved. Although I had to serialize calls to `ensureAllPartitionsConsuming` since there is a possibility of API call running concurrently with the periodic task. I think we can improve upon this by having table level locks being acquired within the function. Have added that as a todo since it should not be much of an issue I feel.

saurabhd336 replied:
Makes sense. TableName level locks won't help much in that case. But synchronizing the method is still needed? To address the, albeit rare, possibility of the API being hit on the lead controller, while the periodic task is running.
Since its a background periodic task, the overhead of the syncronized function shouldn't be much I feel.. Any suggestions on this?

npawar replied:
synchronization is still needed for same controller yes.. but the other case is more concerning. for instance, on the 2 controllers (1 via periodic task thread other via resumeConsumption thread), we could enter setupNewPartitionGroup method at once, and end up with 2 zk metadata and 2 CONSUMING segments. Many more such cases could come up.
Wondering now, if adding a back door entry to this method from any controller is a good idea. This won't happen if only periodic task is allowed to execute that method (be it scheduled or manual). So 1) should this flag just be part of manual periodic task trigger options and then 2) this API internally just invokes the periodic task

sajjad-moradi replied:
I agree. If this method is only called via the periodic task - RealtimeSegmentValidationManager - then we could avoid these potential race problems. The new flag needs to be wired to the periodic task though. Looked at the code for manually running the periodic task. `PinotControllerPeriodicTaskRestletResource.runPeriodicTask()` can be refactored to take task parameters...

saurabhd336 replied:
I was trying to go through helix's `updateIdealState` function, which gets called by `ensureAllPartitionsConsuming`, and looks like the idealState update has support for handling concurrent writes (CAS like implementation). So we may not need syncronization afterall? Even if two threads / processes try to concurrently update the state, it'll get serialized and hence will be safe.
 CC: @Jackie-Jiang for more inputs

sajjad-moradi replied:
@saurabhd336 (cc @Jackie-Jiang) If it was just ideal state changes, what you're saying would be fine, but `ensureAllPartitionsConsuming` method also adds segmentZkMetadata and that part is not serialized!
Manually kicking off RealtimeSegmentValidation job takes care of both ideal states and segZkMetadata changes as it serializes running of the whole job so that only one validation job is running at any given time.

Jackie-Jiang replied:
@sajjad-moradi It should be okay because we always follow the same sequence of operation as the segment commit protocol:
```
Step 1: Update PROPERTYSTORE to change the old segment metadata status to DONE
Step 2: Update PROPERTYSTORE to create the new segment metadata with status IN_PROGRESS
Step 3: Update IDEALSTATES to include new segment in CONSUMING state, and change old segment to ONLINE state.
```
The operation itself is idempotent. Given any intermediate state, it will return the same result.

If the operation is not idempotent, then we also need to handle the synchronization across multiple controllers.

sajjad-moradi replied:
Step 2 is not idempotent. Because of the timestamp in the new segment name, everytime step 2 gets executed, we'll get a new segment name which leads to new segZkMetadata:
```java
      long newSegmentCreationTimeMs = getCurrentTimeMs();
      LLCSegmentName newLLCSegment = new LLCSegmentName(rawTableName, committingSegmentPartitionGroupId,
          committingLLCSegment.getSequenceNumber() + 1, newSegmentCreationTimeMs);
      createNewSegmentZKMetadata(tableConfig, streamConfig, newLLCSegment, newSegmentCreationTimeMs,
          committingSegmentDescriptor, committingSegmentZKMetadata, instancePartitions, numPartitionGroups, numReplicas,
          newPartitionGroupMetadataList);

```

Jackie-Jiang replied:
Good point. So it can potentially lead to race condition if 2 runs happen at exactly the same time, and both create the new segment ZK metadata. One of the segment ZK metadata will be picked and continue, the other will be left orphan. 
@saurabhd336 Let's change `ensureAllPartitionsConsuming` to `synchronized`, and add a TODO describing this potential race condition. We should probably clean up the orphan segment ZK metadata if it is not in the ideal state. This can be handled in a separate PR.

sajjad-moradi replied:
We don't need to do so if we go with manually kicking RealtimeSegmentValidation job as Neha suggested. The code in `PinotControllerPeriodicTaskRestletResource` can be modified a bit to address to achieve this.

saurabhd336 replied:
@sajjad-moradi Not sure I understand 'manually kicking RealtimeSegmentValidation job'. That still means `ensureAllPartitionsConsuming` can get called concurrently by the API thread and the scheduled thread?

npawar replied:
@saurabhd336 just think of the periodic task. There's an API to invoke periodic task manually (look for Periodic Task tab in swagger). Whether it runs by schedule, or by manual trigger, periodic task can only run on lead controller of a table, so that eliminates across controller clashes. As for same controller, if you look at BasePeriodicTask, you'll see it doesn't let more than one task to run at once.

So the suggestion is, rely on this mechanism as we already know this works and prevents clashes. One way to rely on this mechanism is by introducing a query param to the runPeriodicTask API with our extra config and pass it to the periodic task execution. Another way could be to keep your new API, but again use same mechanism happening in runPeriodicTask API to trigger the task.

does this help clarify? We can chat on the slack channel if need be.

saurabhd336 replied:
I did not have full context on the periodic task execution via API. But now I've updated the PR with
1) Allowing custom params to be passed to on demand task execution via helix messages.
2) Changed the API to use the periodic task execution instead of directly calling the method. 
Please have a look @npawar @Jackie-Jiang @sajjad-moradi

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1385,4 +1386,89 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
+        getPartitionGroupConsumptionStatusList(idealState, streamConfig);
+    // Read the smallest offset when a new partition is detected
+    OffsetCriteria originalOffsetCriteria = streamConfig.getOffsetCriteria();
+    streamConfig.setOffsetCriteria(OffsetCriteria.SMALLEST_OFFSET_CRITERIA);
```
That is indeed the case. This will set startOffset to 0 only for new partitions being added. Existing partitions will get startOffset = latest segment's endOffset

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1385,4 +1386,89 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
+
+    int numPartitionGroups = newPartitionGroupMetadataList.size();
+    Map<Integer, SegmentZKMetadata> partitionToLatestSegment = getLatestSegmentZKMetadataMap(realtimeTableName);
+    for (PartitionGroupMetadata partitionGroupMetadata : newPartitionGroupMetadataList) {
```
Intentional, yes. But we can discuss this offline. Idea was to keep this as a generic resumeConsumption API, which also happens to double up as a fix for deleted segments.

npawar replied:
I see. But it's not going to be a generic resume consumption Api unless all the other cases from ensureAllPartitions are also included. Now we have two ways to resume consumption, both doing only partial cases.

sajjad-moradi replied:
What do you think if we make the following changes to `ensureAllPartitionsConsuming` method :
```java
ensureAllPartitionsConsuming(..., boolean enforcedByAdmin) {
  ...
  if (enforcedByAdmin) {
    // handle the case where a realtime segment is manually deleted 
  }
  ...
}
```
Then this method is called in the resume endpoint with enforcedByAdmin = true and in the periodic task with enforcedByAdmin = false.
Later if we have more cases that require admin intervention, we can add them to the `if (enforcedByAdmin) {...}` block.

npawar replied:
this is a great idea. +1 on doing it this way

saurabhd336 replied:
Ack. will move this logic into `ensureAllPartitionsConsuming`
In that case, we'd be creating segments for new partition groups as well @npawar ?

npawar replied:
imo this is okay, as we have all logic in one place and we're not just creating new partition groups but running the whole flow with all the cases. My main concern was that we were just doing half of the things

saurabhd336 replied:
Ack. This question turned out to be redundant since creating new consuming segments for new partitions is already handled by `ensureAllPartitionsConsuming`. 
All in all, the API does everything `ensureAllPartitionsConsuming` already does, and additionally also creates consuming segments for partitions where CONSUMING segments were deleted.

Jackie-Jiang replied:
IMO, having a single flag of `enforcedByAdmin` can be quite confusing and hard to use because this module (segment manager) should not be aware of the intention of the admin. I'd suggest making the flag specific and self-explained, such as `recreateDeletedConsumingSegment`. In the future if we want to add other checks, we can introduce other flags, or wrap the flags into a options class, but each flag should be specific. @sajjad-moradi @npawar What's your opinion on this suggestion?

npawar replied:
my thought was also of a generic name like "fixAllCases" or "fixManualErrors" . But i see your point, self-explanatory name would be a good start and introduce more as and when needed.

sajjad-moradi replied:
That's a valid point. We can start with this specific flag now and later on, when we add more cases, we rename the flag to be more generic or add more options.

saurabhd336 replied:
Ack. Renaming the flag. Wrapping these flags into a Config / Props class eventually sounds good to me.

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,52 @@
+
+@Api(tags = Constants.TABLE_TAG)
+@Path("/")
+public class PinotSegmentStateResource {
```
Ack

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,52 @@
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Resume the consumption of a realtime table",
+      notes = "Resume the consumption of a realtime table")
+  public String resumeRealtimeTableConsumption(
```
Ack

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java:
```diff
@@ -0,0 +1,52 @@
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName) {
+    _pinotLLCRealtimeSegmentManager.resumeRealtimeTableConsumption(tableName);
+    // TODO(saurabh): Change this
+    return "{\"status\": \"SUCCESS\"}";
```
Ack

saurabhd336 commented:
@npawar @Jackie-Jiang @sajjad-moradi I've updated the PR as per the comments. Please have a look.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -862,7 +863,9 @@ private Map<Integer, SegmentZKMetadata> getLatestSegmentZKMetadataMap(String rea
    */
-  public void ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelStreamConfig streamConfig) {
+  // TODO(saurabh) : We can reduce critical section by having tableName level locks?
+  public synchronized void ensureAllPartitionsConsuming(TableConfig tableConfig,
```
Let's update the javadoc for this method to include the new scenario:
```
If the consuming segment is deleted:
Check whether there are segments in the PROPERTYSTORE with status DONE, but no new segment in status IN_PROGRESS, and the state for the latest segment in the IDEALSTATE is ONLINE
```
(Note that this is very similar to the failure between step-1 and step-2, the only difference is the state in the ideal state)

saurabhd336 replied:
> It should work without synchronization because it is performing a check and write

Wasn't aware of this. Looks like helix itself guarantees concurrent updates are handeled safely. Removed synchronization. 



> Let's update the javadoc for this method to include the new scenario:
> 
> ```
> If the consuming segment is deleted:
> Check whether there are segments in the PROPERTYSTORE with status DONE, but no new segment in status IN_PROGRESS, and the state for the latest segment in the IDEALSTATE is ONLINE
> ```
> 
> (Note that this is very similar to the failure between step-1 and step-2, the only difference is the state in the ideal state)

Ack. Added.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1069,6 +1073,10 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
     //       and restart consumption from the same offset (if possible) or a newer offset (if realtime stream does
     //       not have the same offset).
     //       In latter case, report data loss.
+    // 4. There are no segments in the table's idealstate in CONSUMING state,
```
The new scenario is very similar to 1, so we can keep them together:
We first check if the `latestSegmentZKMetadata` is `DONE`, then check if `instanceStateMap` contains `CONSUMING` segment.

saurabhd336 replied:
Ack

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -862,7 +863,9 @@ private Map<Integer, SegmentZKMetadata> getLatestSegmentZKMetadataMap(String rea
-  public void ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelStreamConfig streamConfig) {
+  // TODO(saurabh) : We can reduce critical section by having tableName level locks?
+  public synchronized void ensureAllPartitionsConsuming(TableConfig tableConfig,
+      PartitionLevelStreamConfig streamConfig, boolean enforcedByAdmin) {
```
The idea here is to (possibly) add more such cases that'll need to be handled conditionally, exclusively via the API, or other callers.
https://github.com/apache/pinot/pull/8663#discussion_r872885669

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1147,13 +1155,34 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
+              // add it as CONSUMING segment to IDEALSTATE. Currently, only enabled via /resumeConsumption API
+
+              if (enforcedByAdmin) {
+                LLCSegmentName newLLCSegmentName = getNextLLCSegmentName(latestLLCSegmentName, currentTimeMs);
+                StreamPartitionMsgOffset startOffset = offsetFactory.create(latestSegmentZKMetadata.getEndOffset());
+                StreamPartitionMsgOffset partitionGroupSmallestOffset =
+                    getPartitionGroupSmallestOffset(streamConfig, partitionGroupId);
+
+                // Start offset must be higher than the start offset of the stream
+                if (partitionGroupSmallestOffset.compareTo(startOffset) > 0) {
+                  LOGGER.error("Data lost from offset: {} to: {} for partition: {} of table: {}", startOffset,
+                      partitionGroupSmallestOffset, partitionGroupId, realtimeTableName);
+                  _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.LLC_STREAM_DATA_LOSS, 1L);
+                  startOffset = partitionGroupSmallestOffset;
+                }
+                CommittingSegmentDescriptor committingSegmentDescriptor =
+                    new CommittingSegmentDescriptor(null, startOffset.toString(), 0);
+                createNewSegmentZKMetadata(tableConfig, streamConfig, newLLCSegmentName, currentTimeMs,
+                    committingSegmentDescriptor, latestSegmentZKMetadata, instancePartitions, numPartitions,
+                    numReplicas, newPartitionGroupMetadataList);
+                String newSegmentName = newLLCSegmentName.getSegmentName();
+                updateInstanceStatesForNewConsumingSegment(instanceStatesMap, null, newSegmentName, segmentAssignment,
+                    instancePartitionsMap);
```
Added a new function for this

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1121,39 +1131,23 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
+                segmentAssignment, instancePartitionsMap, true);
           } else {
             if (newPartitionGroupSet.contains(partitionGroupId)) {
               // If we get here, that means in IdealState, the latest segment has no CONSUMING replicas, but has
```
```suggestion
              // If we get here, that means in IdealState, the latest segment has all replicas ONLINE
```

Jackie-Jiang replied:
Let's add some checks before fixing the segment:
- latestSegmentZKMetadata.getStatus() == Status.DONE
- isAllInstancesInState(instanceStateMap, SegmentStateModel.ONLINE)

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1211,6 +1205,43 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
     return idealState;
   }
 
+  private void createNewConsumingSegment(TableConfig tableConfig, PartitionLevelStreamConfig streamConfig,
```
Let's pass in the `startOffset` of the new segment, instead of `offsetFactory` and `isLatestSegmentOffline`

saurabhd336 replied:
Ack

Jackie-Jiang reviewed (commented):
Mostly good

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1061,6 +1068,9 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
+    // 2. The latest metadata is in DONE state, but the idealstate has no segment in CONSUMING state.
+    //    a. Create metadata for new IN_PROGRESS segment with startOffset set to latest segments' end offset.
+    //    b. Add the newly created segment to idealstate with segment state set to CONSUMING.
     // 2. The latest metadata is IN_PROGRESS, but segment is not there in idealstate.
```
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1121,39 +1132,28 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
-              // such state.
-              LOGGER
-                  .error("Got unexpected instance state map: {} for segment: {}", instanceStateMap, latestSegmentName);
+              // If we get here, that means in IdealState, the latest segment has all replicas ONLINE, but no
+              // CONSUMING segments.
```
(nit)
```suggestion
              // If we get here, that means in IdealState, the latest segment has all replicas ONLINE.
```

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1121,39 +1132,28 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
+              // Create a new IN_PROGRESS segment in PROPERTYSTORE,
+              // add it as CONSUMING segment to IDEALSTATE.
+
+              if (recreateDeletedConsumingSegment && Status.DONE.equals(latestSegmentZKMetadata.getStatus())
+                  && isAllInstancesInState(instanceStateMap, SegmentStateModel.ONLINE)
+              ) {
+                StreamPartitionMsgOffset startOffset = offsetFactory.create(latestSegmentZKMetadata.getEndOffset());
+                createNewConsumingSegment(tableConfig, streamConfig, latestSegmentZKMetadata, currentTimeMs,
+                    partitionGroupId, newPartitionGroupMetadataList, instancePartitions,
+                    instanceStatesMap, segmentAssignment, instancePartitionsMap, startOffset);
+              } else {
+                LOGGER.error("Got unexpected instance state map: {} for segment: {}",
+                    instanceStateMap, latestSegmentName);
+              }
```
(code style) Let's reformat this part

saurabhd336 replied:
Ack

Jackie-Jiang reviewed (approved):
LGTM. @npawar @sajjad-moradi Can you please also take another look

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/messages/RunPeriodicTaskMessage.java:
```diff
@@ -32,20 +32,23 @@ public class RunPeriodicTaskMessage extends Message {
   private static final String PERIODIC_TASK_REQUEST_ID = "requestId";
   private static final String PERIODIC_TASK_NAME_KEY = "taskName";
   private static final String TABLE_NAME_WITH_TYPE_KEY = "tableNameWithType";
+  private static final String TASK_PARAMS = "taskParams";
```
Suggest naming it "properties" to be consistent with the name in `RunPeriodicTaskMessageHandler`

saurabhd336 replied:
Ack. Renamed to 'taskProperties'

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/ControllerUserDefinedMessageHandlerFactory.java:
```diff
@@ -109,6 +113,10 @@ private static Properties createTaskProperties(String periodicTaskRequestId, Str
         periodicTaskParameters.setProperty(PeriodicTask.PROPERTY_KEY_TABLE_NAME, tableNameWithType);
       }
 
+      if (taskParamsJson != null) {
```
Here we can write the key-value pairs of the map into the `Properties`

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRealtimeTableResource.java:
```diff
@@ -0,0 +1,99 @@
+    String periodicTaskRequestId = "api-" + UUID.randomUUID().toString().substring(0, 8);
+
+    LOGGER.info("[TaskRequestId: {}] Sending periodic task message to all controllers for running task {} against {}.",
+        periodicTaskRequestId, "RealtimeSegmentValidationManager", " table '" + tableNameWithType + "'");
```
(minor) Make `"RealtimeSegmentValidationManager"` a constant

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRealtimeTableResource.java:
```diff
@@ -0,0 +1,99 @@
+    // with this PeriodicTask's execution.
+    String periodicTaskRequestId = "api-" + UUID.randomUUID().toString().substring(0, 8);
+
+    LOGGER.info("[TaskRequestId: {}] Sending periodic task message to all controllers for running task {} against {}.",
```
Suggest making it more explicit that this is a resume consumption request

saurabhd336 replied:
I've added a new response pojo

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java:
```diff
@@ -83,7 +85,7 @@ protected final void runTask(Properties periodicTaskProperties) {
 
       if (!tablesToProcess.isEmpty()) {
-        processTables(tablesToProcess);
+        processTables(tablesToProcess, taskParamsJson);
```
Suggest passing the whole `Properties` into the `processTables` to be more flexible

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java:
```diff
@@ -103,10 +105,16 @@ public final ControllerMetrics getControllerMetrics() {
-    C context = preprocess();
+    C context;
+    try {
+      context = preprocess(taskParamsJson);
```
Same here, passing the whole `Properties` into `preprocess`

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java:
```diff
@@ -103,10 +105,16 @@ public final ControllerMetrics getControllerMetrics() {
+    C context;
+    try {
+      context = preprocess(taskParamsJson);
+    } catch (IOException e) {
```
We shouldn't need to catch this exception when the key-values are already put into the properties

saurabhd336 replied:
Moving to Properties instead of json makes `preprocess` not throw exceptions. Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:
```diff
@@ -172,8 +183,13 @@ public void cleanUpTask() {
     _validationMetrics.unregisterAllMetrics();
   }
 
+  public static final class RealtimeSegmentValidationTaskParameters {
```
This is already an inner class of `RealtimeSegmentValidationManager`, so no need to have a separate wrapper class for the parameters. Suggest directly adding `_recreateDeletedConsumingSegment` into the `Context` class

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRealtimeTableResource.java:
```diff
@@ -0,0 +1,99 @@
+    // Generate an id for this request by taking first eight characters of a randomly generated UUID. This request id
+    // is returned to the user and also appended to log messages so that user can locate all log messages associated
+    // with this PeriodicTask's execution.
+    String periodicTaskRequestId = "api-" + UUID.randomUUID().toString().substring(0, 8);
```
This part is almost identical to the `PinotControllerPeriodicTaskRestletResource`, let's add a TODO here to extract the common part in the future

saurabhd336 replied:
Ack

saurabhd336 replied:
Ack. PinotHelixResourceManager seemed like the most natural place for the method to reside. Have moved it there and changed both the classes to use it. Please check.

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -855,14 +855,19 @@ private Map<Integer, SegmentZKMetadata> getLatestSegmentZKMetadataMap(String rea
    * idealState.
    * If so, it should create the new segments in idealState.
    *
+   * If the consuming segment is deleted:
```
Mention here that this happens only if the special flag  is set

saurabhd336 replied:
Ack

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1121,39 +1131,26 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
           // 3. we should never end up with some replicas ONLINE and some OFFLINE.
```
remove the "Resolved in " from point number 2, now that your PR handles this. I think that comment was making some assumptions that this will never happen after correcting the bug in that 2980 PR

saurabhd336 replied:
Ack

saurabhd336 commented on pinot-common/src/main/java/org/apache/pinot/common/messages/RunPeriodicTaskMessage.java:
```diff
@@ -32,20 +32,23 @@ public class RunPeriodicTaskMessage extends Message {
    */
-  public RunPeriodicTaskMessage(String taskRequestId, String periodicTaskName, String tableNameWithType) {
+  public RunPeriodicTaskMessage(String taskRequestId, String periodicTaskName, String tableNameWithType,
+      String taskParamsJson) {
```
Ack

sajjad-moradi commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1117,43 +1129,30 @@ IdealState ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelS
-              LOGGER
-                  .error("Got unexpected instance state map: {} for segment: {}", instanceStateMap, latestSegmentName);
+              // If we get here, that means in IdealState, the latest segment has all replicas ONLINE.
+              // Create a new IN_PROGRESS segment in PROPERTYSTORE,
+              // add it as CONSUMING segment to IDEALSTATE.
```
Move this to the next if statement.

saurabhd336 replied:
Ack

sajjad-moradi reviewed (approved):
LGTM, with a few minor comments.

sajjad-moradi commented:
Please add a `Testing Done` section to the PR description that describes your manual testing. It should cover two scenarios:
- Delete a consuming segment using segment delete endpoint (which deletes both IS and segZkMetadata) and then call resume endpoint
- Manually delete IS entries for a consuming segment in ZK and then call resume endpoint

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:
```diff
@@ -79,6 +82,12 @@ protected Context preprocess() {
       _lastSegmentLevelValidationRunTimeMs = currentTimeMs;
     }
+
+    if (periodicTaskProperties.containsKey(RECREATE_DELETED_CONSUMING_SEGMENT_KEY)) {
+        context._recreateDeletedConsumingSegment =
+            Boolean.parseBoolean(periodicTaskProperties.getProperty(RECREATE_DELETED_CONSUMING_SEGMENT_KEY));
+    }
```
(minor) Can be simplified
```suggestion
      context._recreateDeletedConsumingSegment =
          Boolean.parseBoolean(periodicTaskProperties.getProperty(RECREATE_DELETED_CONSUMING_SEGMENT_KEY));
```

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/messages/RunPeriodicTaskMessage.java:
```diff
@@ -63,4 +70,8 @@ public String getPeriodicTaskName() {
     return getRecord().getSimpleField(TABLE_NAME_WITH_TYPE_KEY);
   }
+
+  public Map<String, String> getTaskParams() {
```
```suggestion
  public Map<String, String> getTaskProperties() {
```

saurabhd336 replied:
Ack

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/PeriodicTask.java:
```diff
@@ -32,6 +32,7 @@ public interface PeriodicTask extends Runnable {
   // PeriodicTask objects may take a {@link Properties} object. Define all the keys property keys here.
   String PROPERTY_KEY_REQUEST_ID = "requestId";
   String PROPERTY_KEY_TABLE_NAME = "tableNameWithType";
+  String TASK_PARAMS_JSON = "taskParams";
```
Revert

saurabhd336 replied:
Ack

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -855,14 +855,21 @@ private Map<Integer, SegmentZKMetadata> getLatestSegmentZKMetadataMap(String rea
+   * Check whether there are segments in the PROPERTYSTORE with status DONE, but no new segment in status
+   * IN_PROGRESS, and the state for the latest segment in the IDEALSTATE is ONLINE.
+   * If so, it should create a new CONSUMING segment for the partition.
+   * (this operation is done only if @param recreateDeletedConsumingSegment is set to true)
```
Ack

saurabhd336 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -855,14 +855,21 @@ private Map<Integer, SegmentZKMetadata> getLatestSegmentZKMetadataMap(String rea
    * If the controller fails after step-3, we are fine because the idealState has the new segments.
    * If the controller fails before step-1, the server will see this as an upload failure, and will re-try.
    * @param tableConfig
    *
```
Ack
